readthedocs/proxito/middleware.py

```diff
@@ -293,19 +297,14 @@ def process_request(self, request):  # noqa
             )
             return redirect(final_url)
 
+        project = ret
         log.debug(
             'Proxito Project.',
-            project_slug=ret,
+            project_slug=project.slug,
         )
 
         # Otherwise set the slug on the request
-        request.host_project_slug = request.slug = ret
-
-        try:
-            project = Project.objects.get(slug=request.host_project_slug)
-        except Project.DoesNotExist:
-            log.debug("No host_project_slug set on project")
-            return None
+        request.host_project_slug = request.slug = project.slug
```

Comment:
Can we avoid setting "random" attributes to the `request` object and use _one and only one_ field that it's pretty well defined as a dataclass? Similar to what you've done with `UpdateDocsTask.data` attribute being a `TaskData` dataclass.

I'm imaging something like:

```python
@dataclass(slots=True)
class RequestMetadata:
  host_project_slug: str = None
  slug: str = None
  rtdheader: bool = False
  subdomain: bool = False
  cname: bool = False
  canonicalize: bool = False
```

and then using it as:

```python
request.readthedocs = RequestMetadata(slug=project.slug, ...)
```

Reply:
This is the idea when using the new proxito implementation, I'm not refactoring that part right now.